sockeye/inference.py

```diff
@@ -631,11 +644,51 @@ def _make_result(self,
                                 translation=target_string,
                                 tokens=target_tokens,
                                 attention_matrix=attention_matrix,
-                                score=neg_logprob)
+                                score=translation.score)
+
+    def _concat_translations(self, translations: List[Translation]) -> Translation:
+        """
+        Combine translations through concatenation.
+
+        :param translations: A list of translations (sequence, attention_matrix), score and length.
+        :return: A concatenation if the translations with a score.
+        """
+        # Concatenation of all target ids without BOS and EOS
+        target_ids = list(itertools.chain.from_iterable(translation.target_ids[1:-1] for translation in translations))
+        # Adding back the EOS symbol
+        target_ids = [self.vocab_target[C.BOS_SYMBOL]] + target_ids + [self.vocab_target[C.EOS_SYMBOL]]
+
+        # Combine attention matrices:
+        attention_matrices = [t.attention_matrix for t in translations]
+        attention_shapes = [attention_matrix.shape for attention_matrix in attention_matrices]
+        # We remove EOS and BOS from all matrices except for the first and last
+        combined_shape_target = sum(shape_t - 2 for shape_t, _ in attention_shapes) + 2
+        # The source is kept unchanged
+        combined_shape_source = sum(shape_s for _, shape_s in attention_shapes)
+
+        attention_matrix_combined = np.zeros((combined_shape_target, combined_shape_source))
+
+        # We start at position 1 as position 0 is for the BOS, which is kept zero
+        pos_t, pos_s = 1, 0
+        for attention_matrix, (len_t, len_s) in zip(attention_matrices, attention_shapes):
+            if attention_matrix is attention_matrices[-1]:
+                # Alignment without BOS but with EOS
+                attention_matrix_combined[pos_t:pos_t + len_t-1, pos_s:pos_s + len_s] = attention_matrix[1:,:]
+            else:
+                # Alignment without neither BOS nor EOS
+                attention_matrix_combined[pos_t:pos_t + len_t-2, pos_s:pos_s + len_s] = attention_matrix[1:-1,:]
+                pos_t += len_t-2
+                pos_s += len_s
+
+        # Unnormalize + sum and renormalize:
+        score = sum(translation.score * self.length_penalty(len(translation.target_ids))
+                    for translation in translations)
+        score = score / self.length_penalty(len(target_ids))
```

Comment:
We could be clearer by keeping sufficient statistics for scores and length penalty, effectively our feature set.  Each `Translation` could have `Scores` with clearly defined sum and eval operations.  That can probably wait, but we could add a TODO to refactor this part if we add more score components.

Reply:
yeah, that would make a lot of sense.

Reply:
However, that would require yet another set of variables we need to run sort/take operations on, which tend to be a bit slow. The inplace renormalization is probably faster.

Reply:
Instead of duplicating a lot of EOS/BOS removal logic and handling attention matrices, would it make more sense to move concatenation of translations AFTER the make_result calls? make_result is already handling this stuff and I am afraid that we will add too many places that somehow deal with stripping special symbols from the output.

Reply:
in make_result we already concatenate to strings, which makes the concat more difficult. I wanted to write the concat method in a way to produce an output which is always compatible to make_result, in that the translations still start with BOS for both the target_ids and the attention_matrix.